### DISCUSSION TITLE
sqlalchemy-2 handling of freezetime FakeDate

### DISCUSSION THREAD

**Discussion by HawkOnPK** [sqlalchemy/sqlalchemy#11642] [Usage Questions] [answered]
I have a column in a postgres database of type `DATERANGE`:
`validrange = Column(DATERANGE, nullable=False)`
When filtering on items valid for certain dates, the range is checked on inclusion of the date element:
`if isinstance(valid, date): q = q.filter(Userabsencetype.validrange.op('@>')(valid))`
Test cases use `freezegun`, causing date objects to sometimes be of type `FakeDate`.

Previously, with sqlalchemy v1.4, everything worked fine.
Now, after upgrading to sqlalchhemy 2, queries fail for type `FakeDate`, since the sql parameter now receives an additional cast to `::DATERANGE` as seen in the error:
```
(psycopg2.errors.CannotCoerce) cannot cast type date to daterange
(validrange @> '2024-03-01'::date::DATERANG...

[SQL: ...AND (validrange @> %(validrange_1)s::DATERANGE)]
```

The standard `datetime` objects work without fail.
Adding a custom psycopg2 adapter lets me adjust the `::date` cast, but the latter `::DATERANGE` still appears nontheless.

Has anyone an idea, if that is a bug, or maybe can provide a workaround? Thank you.

**Comment by CaselIT** [comment, marked as answer]
Hi,

The left type of the operator is ported to the right side by default in v2.
You can use `sa.literal(value, sa.Date)` or use `sa.type_coerce(value, sa.Date)` to avoid it.

**Reply by CaselIT** [reply]
@zzzeek maybe literal should accept a flag to turn off the type propagation from the left side?

**Reply by HawkOnPK** [reply]
Thank you. Those do fix the problem indeed.
Yet, it is unclear for me, why standard `date` objects do not cause this issue, but the `freezegun.FakeDate` objects do. I assume, that his is caused by the date objects having a known type and thus a special handling, and the FakeDate does not and falls back to a default type cast. Is there an option to inject own type / cast handler in sqlalchemy? 
Edit: Also, which seems to be strange is that it receives a double cast, for both date and daterange.

**Reply by CaselIT** [reply]
> Is there an option to inject own type / cast handler in sqlalchemy?

specifying the type to literal or using type_coerce
